cylc/flow/task_pool.py

```diff
@@ -400,16 +394,16 @@ def release_runahead_tasks(self):
         if self.stop_point and latest_allowed_point > self.stop_point:
             latest_allowed_point = self.stop_point
 
-        release = []
-        for point, itask_id_map in self.main_pool.items():
-            if point <= latest_allowed_point:
-                for itask in itask_id_map.values():
-                    if not itask.state.is_runahead:
-                        continue
-                    release.append(itask)
-        for itask in release:
+        for itask in (
+            itask
+            for point, itask_id_map in self.main_pool.items()
+            for itask in itask_id_map.values()
+            if point <= latest_allowed_point
+            if itask.state.is_runahead
+        ):
```

Comment:
Saves building and destroying intermediate list.

Reply:
Actually this one broke some tests, which did my head in for a bit because I was trying to see how your version gave a different result.  The problem is, the `release_runahead_tasks()` call inside the loop can sometimes result in the dictionary changing size during iteration (because: parent-less tasks get spawned when their previous instance gets released from runahead-limiting). So the intermediate list is necessary here.  I'll insert a comment to explain why, for future reference.

Reply:
(The original can still be improved a bit though).